pest/src/parser_state.rs

```diff
@@ -858,23 +858,67 @@ impl<'i, R: RuleType> ParserState<'i, R> {
     /// assert_eq!(result.unwrap().position().pos(), 2);
     /// ```
     #[inline]
-    pub fn stack_pop(self: Box<Self>) -> ParseResult<Box<Self>> {
-        let result_state = {
-            let string = self.stack
-                .peek()
-                .expect("pop was called on empty stack")
-                .as_str();
-
-            self.match_string(string)
-        };
+    pub fn stack_pop(mut self: Box<Self>) -> ParseResult<Box<Self>> {
+        let string = self.stack
+            .pop()
+            .expect("pop was called on empty stack")
+            .as_str();
+        self.match_string(string)
+    }
 
-        match result_state {
-            Ok(mut state) => {
-                state.stack.pop();
-                Ok(state)
-            }
-            Err(state) => Err(state)
+    /// Matches the full state of the stack.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use pest;
+    /// # #[allow(non_camel_case_types)]
+    /// # #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+    /// enum Rule {}
+    ///
+    /// let input = "aaaa";
+    /// let mut state: Box<pest::ParserState<Rule>> = pest::ParserState::new(input);
+    /// let mut result = state.stack_push( |state| state.match_string("a")).and_then(|state| {
+    ///     state.stack_push( |state| state.match_string("a"))
+    /// }).and_then(|state| state.stack_match_peek());
+    /// assert!(result.is_ok());
+    /// assert_eq!(result.unwrap().position().pos(), 4);
+    /// ```
+    #[inline]
+    pub fn stack_match_peek(self: Box<Self>) -> ParseResult<Box<Self>> {
+        let spans = self.stack.get_state();
+        let string = spans.iter()
+            .map(|s| s.as_str())
+            .collect::<Vec<&str>>()
+            .join("");
+        self.match_string(&string)
+    }
+
+    /// Matches the full state of the stack. This method will clear the stack as it evaluates.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// /// # use pest;
+    /// # #[allow(non_camel_case_types)]
+    /// # #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+    /// enum Rule {}
+    ///
+    /// let input = "aaaa";
+    /// let mut state: Box<pest::ParserState<Rule>> = pest::ParserState::new(input);
+    /// let mut result = state.stack_push( |state| state.match_string("a")).and_then(|state| {
+    ///     state.stack_push( |state| state.match_string("a"))
+    /// }).and_then(|state| state.stack_match_peek());
+    /// assert!(result.is_ok());
+    /// assert_eq!(result.unwrap().position().pos(), 4);
+    /// ```
+    #[inline]
+    pub fn stack_match_pop(mut self: Box<Self>) -> ParseResult<Box<Self>> {
+        let mut strings = Vec::new();
```

Comment:
Maybe there one could define `Stack::get_state_mut`, then apply [`Vec::drain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain), then `Iterator::all`?

Reply:
Drain won't work because we need to keep track of all the mutations to the stack which we get through the `pop` call.